### PR TITLE
Fix offset __inits__, apply_index dtypes

### PR DESCRIPTION
**Pull request by jbrockmendel**
This PR does two main things:

1) Change the default kwargs for offset `__init__` methods so that the defaults are valid; ATM `WeekOfMonth()` and `LastWeekOfMonth()` will both raise  `ValueError`.

2) ATM `Week.apply_index` adds an int64 array to a `DatetimeIndex`.  This happens to have the desired output because the int64 array represents nanoseconds.  This explicitly casts the array to `timedelta64[ns]` to make this explicit.  <b>This is a necessary step towards fixing #19123.</b>

In addition there are two small pieces of refactoring:

3) Remove `liboffsets.EndMixin` since it is only mixed in to `Week`.

4) Use `liboffsets.roll_convention` to de-duplicate some code in `liboffsets.shift_months`.

- [ ] closes #xxxx
- [ ] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Comment by jbrockmendel** [comment]
Travis looks like a moto problem.

File: pandas/tseries/offsets.py [hunk 1]
```diff
@@ -1226,7 +1225,8 @@ def _get_roll(self, i, before_day_of_month, after_day_of_month):
         return roll
 
     def _apply_index_days(self, i, roll):
-        i += (roll % 2) * Timedelta(days=self.day_of_month).value
+        nanos = (roll % 2) * Timedelta(days=self.day_of_month).value
```
Comment (jreback) [inline pandas/tseries/offsets.py]: I would rather

```
n = Timedelta((roll % 2) * Timedelta(days=self.day_of_month).value)
return i + n + Timedelta(days-1)
```

File: pandas/tseries/offsets.py [hunk 2]
```diff
@@ -1271,13 +1271,14 @@ def _get_roll(self, i, before_day_of_month, after_day_of_month):
         return roll
 
     def _apply_index_days(self, i, roll):
-        return i + (roll % 2) * Timedelta(days=self.day_of_month - 1).value
+        nanos = (roll % 2) * Timedelta(days=self.day_of_month - 1).value
```
Comment (jreback) [inline pandas/tseries/offsets.py]: same

File: pandas/tseries/offsets.py [hunk 3]
```diff
@@ -1394,7 +1411,7 @@ class WeekOfMonth(_WeekOfMonthMixin, DateOffset):
     _prefix = 'WOM'
     _adjust_dst = True
 
-    def __init__(self, n=1, normalize=False, week=None, weekday=None):
```
Comment (jreback) [inline pandas/tseries/offsets.py]: this doesn't break anything???

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tseries/offsets.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/tseries/offsets.py]: The default Kmarts raise ATM.

**Comment by jbrockmendel** [comment]
Brief interlude to gripe: pytest makes it difficult to run tests in isolation and very difficult to recreate test failures interactively.  In troubleshooting the current Travis failure I could get recreate the failure when running pytest locally but not when interactively running the same code.  As it turns out, the tests in `test_dt64_with_DateOffsets` are _not_ independent of each other.  Still need to figure out how and why...

File: pandas/tseries/offsets.py (hunk 1 above)
Comment (jreback) [reply pandas/tseries/offsets.py]: then pls add a doc-string (same below)

File: pandas/tseries/offsets.py (hunk 3 above)
Comment (jreback) [reply pandas/tseries/offsets.py]: where is the test for this? why isn't it raising currently?

File: pandas/tseries/offsets.py [hunk 4]
```diff
@@ -1332,7 +1333,24 @@ def apply_index(self, i):
             return ((i.to_period('W') + self.n).to_timestamp() +
                     i.to_perioddelta('W'))
         else:
-            return self._end_apply_index(i, self.freqstr)
+            return self._end_apply_index(i)
+
+    def _end_apply_index(self, dtindex):
+        """Offsets index to end of Period frequency"""
```
Comment (jreback) [inline pandas/tseries/offsets.py]: doc-string

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tseries/offsets.py (hunk 4 above)
Comment (jbrockmendel) [reply pandas/tseries/offsets.py]: Will do.

**Comment by jbrockmendel** [comment]
> where is the test for this? why isn't it raising currently?

Until a few dozen PRs ago all that was passed to `__init__` was `n=1, normalize=False, **kwargs`.  The tests had to explicitly pass the relevant kwargs or else you'd get a `KeyError`.  So there aren't any tests that use the defaults.

**Comment by jreback** [comment]
> Until a few dozen PRs ago all that was passed to __init__ was n=1, normalize=False, **kwargs. The tests had to explicitly pass the relevant kwargs or else you'd get a KeyError. So there aren't any tests that use the defaults.

ok can you add some pls.

**Comment by jreback** [comment]
since ``pd.offsets.WeekOfMonth()`` now works, pls add a note in bug fixes for whatsnew. otherwise lgtm. ping on green.

**Comment by jbrockmendel** [comment]
ping

File: doc/source/whatsnew/v0.23.0.txt [hunk 5]
```diff
@@ -474,3 +474,4 @@ Other
 
 - Improved error message when attempting to use a Python keyword as an identifier in a ``numexpr`` backed query (:issue:`18221`)
 - :func:`Timestamp.replace` will now handle Daylight Savings transitions gracefully (:issue:`18319`)
+- Bug in :class:`WeekOfMonth` and :class:`LastWeekOfMonth` where default keyword arguments for constructor raised ``ValueError`` (:issue:`19142`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.23.0.txt]: pls move this and the one above to COnversion, don't put things Other unless they obviously don't fit elsewhere. You *could* make a TimesSeries section if you want (and move appropriate things), but move this for now.

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.23.0.txt (hunk 5 above)
Comment (jbrockmendel) [reply doc/source/whatsnew/v0.23.0.txt]: OK.  For future reference, what heuristic should I be using that maps `WeekOfMonth.__init__ ValueError` to "conversion"?

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks!

**Pull request merged**